From 1a322eb941b5e35fac9ac3f8e33c780821413c90 Mon Sep 17 00:00:00 2001
From: Xie XiuQi <xiexiuqi@huawei.com>
Date: Sat, 25 Jun 2022 01:25:09 +0800
Subject: [PATCH 887/905] [Huawei] Backport legacy percpu-rwsem locking for
 cgroup

Offering: HULK
hulk inclusion
category: feature
bugzilla: NA

Reference: https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/kernel/locking/percpu-rwsem.c?id=7111951b8d4973bda27ff663f2cf18b663d15b48

--------------------------------

Port legacy percpu-rwsem locking for cgroup from v5.6

[XQ: rename percpu-rwsem to legacy to legacy-percpu-rwsem]
Signed-off-by: Xie XiuQi <xiexiuqi@huawei.com>
Signed-off-by: Jian Cheng <cj.chengjian@huawei.com>
Signed-off-by: Wei Li <liwei391@huawei.com>
---
 include/linux/legacy-percpu-rwsem.h  | 149 +++++++++++++++++++
 kernel/locking/Makefile              |   2 +-
 kernel/locking/legacy-percpu-rwsem.c | 206 +++++++++++++++++++++++++++
 3 files changed, 356 insertions(+), 1 deletion(-)
 create mode 100644 include/linux/legacy-percpu-rwsem.h
 create mode 100644 kernel/locking/legacy-percpu-rwsem.c

diff --git a/include/linux/legacy-percpu-rwsem.h b/include/linux/legacy-percpu-rwsem.h
new file mode 100644
index 000000000000..2b011fb9a44e
--- /dev/null
+++ b/include/linux/legacy-percpu-rwsem.h
@@ -0,0 +1,149 @@
+/* SPDX-License-Identifier: GPL-2.0 */
+#ifndef _LINUX_LEGACY_PERCPU_RWSEM_H
+#define _LINUX_LEGACY_PERCPU_RWSEM_H
+
+#include <linux/atomic.h>
+#include <linux/rwsem.h>
+#include <linux/percpu.h>
+#include <linux/rcuwait.h>
+#include <linux/rcu_sync.h>
+#include <linux/lockdep.h>
+
+struct legacy_percpu_rw_semaphore {
+	struct rcu_sync		rss;
+	unsigned int __percpu	*read_count;
+	struct rw_semaphore	rw_sem; /* slowpath */
+	struct rcuwait          writer; /* blocked writer */
+	int			readers_block;
+#ifdef CONFIG_DEBUG_LOCK_ALLOC
+	struct lockdep_map	dep_map;
+#endif
+};
+
+#ifdef CONFIG_DEBUG_LOCK_ALLOC
+#define __LEGACY_PERCPU_RWSEM_DEP_MAP_INIT(lockname)	.dep_map = { .name = #lockname },
+#else
+#define __LEGACY_PERCPU_RWSEM_DEP_MAP_INIT(lockname)
+#endif
+
+#define __DEFINE_LEGACY_PERCPU_RWSEM(name, is_static)				\
+static DEFINE_PER_CPU(unsigned int, __legacy_percpu_rwsem_rc_##name);		\
+is_static struct legacy_percpu_rw_semaphore name = {				\
+	.rss = __RCU_SYNC_INITIALIZER(name.rss),			\
+	.read_count = &__legacy_percpu_rwsem_rc_##name,			\
+	.rw_sem = __RWSEM_INITIALIZER(name.rw_sem),			\
+	.writer = __RCUWAIT_INITIALIZER(name.writer),			\
+	__PERCPU_RWSEM_DEP_MAP_INIT(name)				\
+}
+
+#define DEFINE_LEGACY_PERCPU_RWSEM(name)		\
+	__DEFINE_LEGACY_PERCPU_RWSEM(name, /* not static */)
+#define DEFINE_STATIC_LEGACY_PERCPU_RWSEM(name)	\
+	__DEFINE_PERCPU_RWSEM(name, static)
+
+extern bool __legacy_percpu_down_read(struct legacy_percpu_rw_semaphore *, bool);
+extern void __legacy_percpu_up_read(struct legacy_percpu_rw_semaphore *);
+
+static inline void legacy_percpu_down_read(struct legacy_percpu_rw_semaphore *sem)
+{
+	might_sleep();
+
+	rwsem_acquire_read(&sem->dep_map, 0, 0, _RET_IP_);
+
+	preempt_disable();
+	/*
+	 * We are in an RCU-sched read-side critical section, so the writer
+	 * cannot both change sem->state from readers_fast and start checking
+	 * counters while we are here. So if we see !sem->state, we know that
+	 * the writer won't be checking until we're past the preempt_enable()
+	 * and that once the synchronize_rcu() is done, the writer will see
+	 * anything we did within this RCU-sched read-size critical section.
+	 */
+	if (likely(rcu_sync_is_idle(&sem->rss)))
+		this_cpu_inc(*sem->read_count);
+	else
+		__legacy_percpu_down_read(sem, false); /* Unconditional memory barrier */
+	/*
+	 * The preempt_enable() prevents the compiler from
+	 * bleeding the critical section out.
+	 */
+	preempt_enable();
+}
+
+static inline bool legacy_percpu_down_read_trylock(struct legacy_percpu_rw_semaphore *sem)
+{
+	bool ret = true;
+
+	preempt_disable();
+	/*
+	 * Same as in percpu_down_read().
+	 */
+	if (likely(rcu_sync_is_idle(&sem->rss)))
+		this_cpu_inc(*sem->read_count);
+	else
+		ret = __legacy_percpu_down_read(sem, true); /* Unconditional memory barrier */
+	preempt_enable();
+	/*
+	 * The barrier() from preempt_enable() prevents the compiler from
+	 * bleeding the critical section out.
+	 */
+
+	if (ret)
+		rwsem_acquire_read(&sem->dep_map, 0, 1, _RET_IP_);
+
+	return ret;
+}
+
+static inline void legacy_percpu_up_read(struct legacy_percpu_rw_semaphore *sem)
+{
+	rwsem_release(&sem->dep_map, _RET_IP_);
+
+	preempt_disable();
+	/*
+	 * Same as in legacy_percpu_down_read().
+	 */
+	if (likely(rcu_sync_is_idle(&sem->rss)))
+		this_cpu_dec(*sem->read_count);
+	else
+		__legacy_percpu_up_read(sem); /* Unconditional memory barrier */
+	preempt_enable();
+}
+
+extern void legacy_percpu_down_write(struct legacy_percpu_rw_semaphore *);
+extern void legacy_percpu_up_write(struct legacy_percpu_rw_semaphore *);
+
+extern int __legacy_percpu_init_rwsem(struct legacy_percpu_rw_semaphore *,
+				const char *, struct lock_class_key *);
+
+extern void legacy_percpu_free_rwsem(struct legacy_percpu_rw_semaphore *);
+
+#define legacy_percpu_init_rwsem(sem)					\
+({								\
+	static struct lock_class_key rwsem_key;			\
+	__legacy_percpu_init_rwsem(sem, #sem, &rwsem_key);		\
+})
+
+#define legacy_percpu_rwsem_is_held(sem)	lockdep_is_held(sem)
+#define legacy_percpu_rwsem_assert_held(sem)	lockdep_assert_held(sem)
+
+static inline void legacy_percpu_rwsem_release(struct legacy_percpu_rw_semaphore *sem,
+					bool read, unsigned long ip)
+{
+	lock_release(&sem->dep_map, ip);
+#ifdef CONFIG_RWSEM_SPIN_ON_OWNER
+	if (!read)
+		atomic_long_set(&sem->rw_sem.owner, RWSEM_OWNER_UNKNOWN);
+#endif
+}
+
+static inline void legacy_percpu_rwsem_acquire(struct legacy_percpu_rw_semaphore *sem,
+					bool read, unsigned long ip)
+{
+	lock_acquire(&sem->dep_map, 0, 1, read, 1, NULL, ip);
+#ifdef CONFIG_RWSEM_SPIN_ON_OWNER
+	if (!read)
+		atomic_long_set(&sem->rw_sem.owner, (long)current);
+#endif
+}
+
+#endif
diff --git a/kernel/locking/Makefile b/kernel/locking/Makefile
index 6d11cfb9b41f..efac4b446306 100644
--- a/kernel/locking/Makefile
+++ b/kernel/locking/Makefile
@@ -3,7 +3,7 @@
 # and is generally not a function of system call inputs.
 KCOV_INSTRUMENT		:= n
 
-obj-y += mutex.o semaphore.o rwsem.o percpu-rwsem.o
+obj-y += mutex.o semaphore.o rwsem.o percpu-rwsem.o legacy-percpu-rwsem.o
 
 # Avoid recursion lockdep -> KCSAN -> ... -> lockdep.
 KCSAN_SANITIZE_lockdep.o := n
diff --git a/kernel/locking/legacy-percpu-rwsem.c b/kernel/locking/legacy-percpu-rwsem.c
new file mode 100644
index 000000000000..594db1fd4274
--- /dev/null
+++ b/kernel/locking/legacy-percpu-rwsem.c
@@ -0,0 +1,206 @@
+// SPDX-License-Identifier: GPL-2.0-only
+#include <linux/atomic.h>
+#include <linux/rwsem.h>
+#include <linux/percpu.h>
+#include <linux/lockdep.h>
+#include <linux/legacy-percpu-rwsem.h>
+#include <linux/rcupdate.h>
+#include <linux/sched.h>
+#include <linux/errno.h>
+
+#include "rwsem.h"
+
+int __legacy_percpu_init_rwsem(struct legacy_percpu_rw_semaphore *sem,
+			const char *name, struct lock_class_key *key)
+{
+	sem->read_count = alloc_percpu(int);
+	if (unlikely(!sem->read_count))
+		return -ENOMEM;
+
+	/* ->rw_sem represents the whole legacy_percpu_rw_semaphore for lockdep */
+	rcu_sync_init(&sem->rss);
+	init_rwsem(&sem->rw_sem);
+	rcuwait_init(&sem->writer);
+	sem->readers_block = 0;
+#ifdef CONFIG_DEBUG_LOCK_ALLOC
+	debug_check_no_locks_freed((void *)sem, sizeof(*sem));
+	lockdep_init_map(&sem->dep_map, name, key, 0);
+#endif
+	return 0;
+}
+
+void legacy_percpu_free_rwsem(struct legacy_percpu_rw_semaphore *sem)
+{
+	/*
+	 * XXX: temporary kludge. The error path in alloc_super()
+	 * assumes that legacy_percpu_free_rwsem() is safe after kzalloc().
+	 */
+	if (!sem->read_count)
+		return;
+
+	rcu_sync_dtor(&sem->rss);
+	free_percpu(sem->read_count);
+	sem->read_count = NULL; /* catch use after free bugs */
+}
+
+static bool __legacy_percpu_down_read_trylock(struct legacy_percpu_rw_semaphore *sem)
+{
+	this_cpu_inc(*sem->read_count);
+
+	/*
+	 * Due to having preemption disabled the decrement happens on
+	 * the same CPU as the increment, avoiding the
+	 * increment-on-one-CPU-and-decrement-on-another problem.
+	 *
+	 * If the reader misses the writer's assignment of readers_block, then
+	 * the writer is guaranteed to see the reader's increment.
+	 *
+	 * Conversely, any readers that increment their sem->read_count after
+	 * the writer looks are guaranteed to see the readers_block value,
+	 * which in turn means that they are guaranteed to immediately
+	 * decrement their sem->read_count, so that it doesn't matter that the
+	 * writer missed them.
+	 */
+
+	smp_mb(); /* A matches D */
+
+	/*
+	 * If !readers_block the critical section starts here, matched by the
+	 * release in legacy_percpu_up_write().
+	 */
+	if (likely(!smp_load_acquire(&sem->readers_block)))
+		return true;
+
+	this_cpu_dec(*sem->read_count);
+
+	/* Prod writer to re-evaluate readers_active_check() */
+	rcuwait_wake_up(&sem->writer);
+
+	return false;
+}
+
+bool __legacy_percpu_down_read(struct legacy_percpu_rw_semaphore *sem, bool try)
+{
+	if (__legacy_percpu_down_read_trylock(sem))
+		return true;
+
+	if (try)
+		return false;
+
+	/*
+	 * We either call schedule() in the wait, or we'll fall through
+	 * and reschedule on the preempt_enable() in percpu_down_read().
+	 */
+	preempt_enable_no_resched();
+
+	/*
+	 * Avoid lockdep for the down/up_read() we already have them.
+	 */
+	__down_read(&sem->rw_sem);
+	this_cpu_inc(*sem->read_count);
+	__up_read(&sem->rw_sem);
+
+	preempt_disable();
+	return true;
+}
+
+void __legacy_percpu_up_read(struct legacy_percpu_rw_semaphore *sem)
+{
+	smp_mb(); /* B matches C */
+	/*
+	 * In other words, if they see our decrement (presumably to aggregate
+	 * zero, as that is the only time it matters) they will also see our
+	 * critical section.
+	 */
+	__this_cpu_dec(*sem->read_count);
+
+	/* Prod writer to recheck readers_active */
+	rcuwait_wake_up(&sem->writer);
+}
+
+#define legacy_per_cpu_sum(var)						\
+({									\
+	typeof(var) __sum = 0;						\
+	int cpu;							\
+	compiletime_assert_atomic_type(__sum);				\
+	for_each_possible_cpu(cpu)					\
+		__sum += per_cpu(var, cpu);				\
+	__sum;								\
+})
+
+/*
+ * Return true if the modular sum of the sem->read_count per-CPU variable is
+ * zero.  If this sum is zero, then it is stable due to the fact that if any
+ * newly arriving readers increment a given counter, they will immediately
+ * decrement that same counter.
+ */
+static bool readers_active_check(struct legacy_percpu_rw_semaphore *sem)
+{
+	if (legacy_per_cpu_sum(*sem->read_count) != 0)
+		return false;
+
+	/*
+	 * If we observed the decrement; ensure we see the entire critical
+	 * section.
+	 */
+
+	smp_mb(); /* C matches B */
+
+	return true;
+}
+
+void legacy_percpu_down_write(struct legacy_percpu_rw_semaphore *sem)
+{
+	rwsem_acquire(&sem->dep_map, 0, 0, _RET_IP_);
+
+	/* Notify readers to take the slow path. */
+	rcu_sync_enter(&sem->rss);
+
+	__down_write(&sem->rw_sem);
+
+	/*
+	 * Notify new readers to block; up until now, and thus throughout the
+	 * longish rcu_sync_enter() above, new readers could still come in.
+	 */
+	WRITE_ONCE(sem->readers_block, 1);
+
+	smp_mb(); /* D matches A */
+
+	/*
+	 * If they don't see our writer of readers_block, then we are
+	 * guaranteed to see their sem->read_count increment, and therefore
+	 * will wait for them.
+	 */
+
+	/* Wait for all now active readers to complete. */
+	rcuwait_wait_event(&sem->writer, readers_active_check(sem), TASK_UNINTERRUPTIBLE);
+}
+
+void legacy_percpu_up_write(struct legacy_percpu_rw_semaphore *sem)
+{
+	rwsem_release(&sem->dep_map, _RET_IP_);
+
+	/*
+	 * Signal the writer is done, no fast path yet.
+	 *
+	 * One reason that we cannot just immediately flip to readers_fast is
+	 * that new readers might fail to see the results of this writer's
+	 * critical section.
+	 *
+	 * Therefore we force it through the slow path which guarantees an
+	 * acquire and thereby guarantees the critical section's consistency.
+	 */
+	smp_store_release(&sem->readers_block, 0);
+
+	/*
+	 * Release the write lock, this will allow readers back in the game.
+	 */
+	__up_write(&sem->rw_sem);
+
+	/*
+	 * Once this completes (at least one RCU-sched grace period hence) the
+	 * reader fast path will be available again. Safe to use outside the
+	 * exclusive write lock because its counting.
+	 */
+	rcu_sync_exit(&sem->rss);
+}
-- 
2.17.1

